pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -291,43 +305,58 @@ private RebalanceResult doRebalance(TableConfig tableConfig, RebalanceConfig reb
               + "rebalance", rebalanceJobId, tableNameWithType), e);
       return new RebalanceResult(rebalanceJobId, RebalanceResult.Status.FAILED,
           "Caught exception while calculating target assignment: " + e, instancePartitionsMap,
-          tierToInstancePartitionsMap, null, preChecksResult);
+          tierToInstancePartitionsMap, null, preChecksResult, null);
     }
 
     boolean segmentAssignmentUnchanged = currentAssignment.equals(targetAssignment);
     LOGGER.info("For rebalanceId: {}, instancePartitionsUnchanged: {}, tierInstancePartitionsUnchanged: {}, "
             + "segmentAssignmentUnchanged: {} for table: {}", rebalanceJobId, instancePartitionsUnchanged,
         tierInstancePartitionsUnchanged, segmentAssignmentUnchanged, tableNameWithType);
 
+    RebalanceSummaryResult summaryResult = null;
+    if (summary) {
+      summaryResult = calculateDryRunSummary(currentAssignment, targetAssignment, tableNameWithType, rebalanceJobId);
+    }
```

Comment:
how about calculating the summary after this whole if-block, so we save the cost from calculating it when the table is already balanced?
```
if (segmentAssignmentUnchanged) {
..
}
```

Reply:
I think it might still be useful to see the summary in case the table being balanced is not expected? wdyt?
I haven't moved it yet, but let me know if you still think it's not worth calculating the summary for this scenario and I can move it down and adjust the tests.

Reply:
done